pkg/operators/common/systemdcache.go

```diff
@@ -0,0 +1,219 @@
+	tickerDuration time.Duration
+
+	useCount      int
+	useCountMutex sync.Mutex
+}
+
+var (
+	systemdCache *SystemdCache
+	systemdOnce  sync.Once
+)
+
+func GetSystemdCache() (*SystemdCache, error) {
+	systemdOnce.Do(func() {
+		systemdCache, err = newSystemdCache(1 * time.Second)
+	})
+	return systemdCache, err
+}
+
+func (cache *SystemdCache) Subscribe(sub SystemdSubInterface, publishOldEntries bool) {
+	logrus.Debugf("Subscribing to systemd cache")
+	cache.subMutex.Lock()
+	cache.subscribers = append(cache.subscribers, sub)
+	cache.subMutex.Unlock()
+
+	// Publish old entries after subscribing
+	// This ensures that the subscriber always has all entries that it needs
+	// It may recieve some entries twice because of that
+	if publishOldEntries {
+		cache.idToUnitname.Range(func(key, value interface{}) bool {
+			sub.AddCgroup(value.(string), key.(uint64))
+			return true
+		})
+	}
+}
+
+func (cache *SystemdCache) Unsubscribe(sub SystemdSubInterface) {
+	cache.subMutex.Lock()
+	defer cache.subMutex.Unlock()
+	for i, s := range cache.subscribers {
+		if s == sub {
+			cache.subscribers = append(cache.subscribers[:i], cache.subscribers[i+1:]...)
+			return
+		}
+	}
+}
+
+func newSystemdCache(tickerDuration time.Duration) (*SystemdCache, error) {
+	dbusConn, err := dbus.NewSystemConnectionContext(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("creating dbus connection: %w", err)
+	}
+
+	return &SystemdCache{
+		dbusConn:       dbusConn,
+		tickerDuration: tickerDuration,
+	}, nil
+}
+
+func (cache *SystemdCache) loop() {
+	for {
+		select {
+		case <-cache.exit:
+			return
+		case <-cache.ticker.C:
+			cache.update()
+		}
+	}
+}
+
+func (cache *SystemdCache) Close() {
+	if cache.exit != nil {
+		close(cache.exit)
+		cache.exit = nil
+	}
+	if cache.ticker != nil {
+		cache.ticker.Stop()
+		cache.ticker = nil
+	}
+}
+
+func (cache *SystemdCache) Start() {
+	cache.useCountMutex.Lock()
+	defer cache.useCountMutex.Unlock()
+
+	// No uses before us, we are the first one
+	if cache.useCount == 0 {
+		cache.update()
+		cache.exit = make(chan struct{})
+		cache.ticker = time.NewTicker(cache.tickerDuration)
+		go cache.loop()
+	}
+	cache.useCount++
+}
+
+func (cache *SystemdCache) Stop() {
+	cache.useCountMutex.Lock()
+	defer cache.useCountMutex.Unlock()
+
+	// We are the last user, stop everything
+	if cache.useCount == 1 {
+		cache.Close()
+	}
+	cache.useCount--
+}
+
+func (cache *SystemdCache) publishAdd(unitName string, id uint64) {
+	cache.subMutex.Lock()
+	defer cache.subMutex.Unlock()
+	for _, updater := range cache.subscribers {
+		updater.AddCgroup(unitName, id)
+	}
+}
+
+func (cache *SystemdCache) publishRemove(unitName string, id uint64) {
+	cache.subMutex.Lock()
+	defer cache.subMutex.Unlock()
+	for _, updater := range cache.subscribers {
+		updater.RemoveCgroup(unitName, id)
+	}
+}
+
+func (cache *SystemdCache) update() {
+	units, err := cache.dbusConn.ListUnitsByPatternsContext(context.TODO(), []string{"active"}, []string{})
```

Comment:
Is it really necessary to ask the whole list every time? Couldn't we subscribe to the changes and forward the updates instead of doing it manually?